### ISSUE TITLE
sort! performance

### ISSUE THREAD

**Issue by clynamen** [JuliaData/DataFrames.jl#1927]
Maybe I am wrong, but it seems that the sort function is slow:

```
using Random
using DataFrames

N = convert(Int, 1e7) 

A = rand(0:100, N, 4)
A = DataFrame(A, [:a, :b, :c, :d])

t_begin = time()
sort!(A, [:b, :d])
println("Sort time was $(time() - t_begin)")                                                                                                                              
```
the dataframe is sorted in ~30s 

The equivalent in pandas instead:

```
import pandas as pd 
from time import time
import numpy as np

N = int(1e7)
A = np.random.randint(0, 1000, size=(N, 4))
A = pd.DataFrame(A, columns=['a', 'b', 'c', 'd'])

begin_t = time() 
A.sort_values(['b', 'd'], inplace=True)
print("Sort time was {}".format(time()-begin_t))
```

takes ~3 seconds

Is this similar to the problem of GroupBy performance?

**Comment by bkamins** [comment]
This is indeed bad, and can be improved. `groupby` runs in around 0.5 second on it and even `by(identity, A, [:b, :d], sort=true)` is much faster. And `sort(A, [:b, :d])` is much faster (around 2 seconds).

I have not worked on the `sort!` code, but maybe @nalimilan can comment if it can be improved. In particular this:
```
julia> @time B = sort(A, [:b, :d]); A .= B;
  2.152137 seconds (40.00 M allocations: 991.792 MiB, 10.17% gc time)
```
sorts `A` in place (though it allocates in the mean time a lot), so there is a room for improvement for sure.

@clynamen - you can use the latter pattern (`sort` + broadcasting assignment) in the mean time if you need performance and in place update of `A`.

**Comment by nalimilan** [comment]
Good catch. I hadn't looked at the sorting code, and it turns out it's completely type-unstable currently. I've filed https://github.com/JuliaData/DataFrames.jl/pull/1934 to specialize on the type of sorting columns, and now it takes less than one second (after the first run with this type).

There may well be other improvements to apply, but this was the obvious one.
